source-oracle/replication.go

```diff
@@ -0,0 +1,680 @@
+	s.errCh = make(chan error)
+	s.decodeCh = make(chan logminerMessage, replicationBufferSize)
+	s.cancel = streamCancel
+
+	eg.Go(func() error {
+		return s.run(streamCtx)
+	})
+	eg.Go(func() error {
+		return s.decodeMessageWorker(streamCtx)
+	})
+
+	go func() {
+		if err := eg.Wait(); err != nil {
+			// Context cancellation typically occurs only in tests, and for test stability
+			// it should be considered a clean shutdown and not necessarily an error.
+			if errors.Is(err, context.Canceled) {
+				err = nil
+			}
+			close(s.events)
+			s.logminerStmt.Close()
+			s.conn.Close()
+			s.errCh <- err
+		}
+	}()
+
+	return nil
+}
+
+const pollInterval = 10 * time.Second
+
+// run is the main loop of the replicationStream which loops message
+// receiving and relaying
+func (s *replicationStream) run(ctx context.Context) error {
+	var poll = time.NewTicker(pollInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-poll.C:
+		}
+
+		var err = s.poll(ctx)
+		if err != nil {
+			// ORA-01013: user requested cancel of current operation
+			// this means a cancellation of context happened, which we don't consider an error
+			// since it should only happen in tests
+			if strings.Contains(err.Error(), "ORA-01013") {
+				continue
+			}
+			return fmt.Errorf("failed to poll messages: %w", err)
+		}
+	}
+}
+
+func (s *replicationStream) poll(ctx context.Context) error {
+	for {
+		var startSCN = s.lastTxnEndSCN
+		var currentSCN int
+		var row = s.conn.QueryRowContext(ctx, "SELECT current_scn FROM V$DATABASE")
+		if err := row.Scan(&currentSCN); err != nil {
+			return fmt.Errorf("fetching current SCN: %w", err)
+		}
+		var endSCN = startSCN + s.db.config.Advanced.IncrementalSCNRange
+
+		if currentSCN < endSCN {
+			endSCN = currentSCN
+		}
+
+		// If redo files have switched, we need to restart and add log files
+		// over again, otherwise we just start a new session with a different SCN range
+		if switched, err := s.redoFileSwitched(ctx); err != nil {
+			return err
+		} else if switched {
+			if err := s.endLogminer(ctx); err != nil {
+				return err
+			} else if err := s.addLogFiles(ctx, startSCN, endSCN); err != nil {
+				return err
+			} else if err := s.startLogminer(ctx, startSCN, endSCN); err != nil {
+				return err
+			}
+		} else {
+			if err := s.startLogminer(ctx, startSCN, endSCN); err != nil {
+				return err
+			}
+		}
+
+		if err := s.receiveMessages(ctx); err != nil {
+			return fmt.Errorf("receive messages: %w", err)
+		}
+
+		// Although the query is inclusive, we do not +1 here since
+		// it is possible for there to be multiple rows with the same SCN
+		// and to be cautious we use the SCN as-is. This may lead to duplicate events being captured
+		// on restart, but it saves us from missing events
+		s.lastTxnEndSCN = endSCN
+
+		s.decodeCh <- logminerMessage{Op: opFlush, SCN: s.lastTxnEndSCN}
+	}
+}
+
+func (s *replicationStream) decodeMessageWorker(ctx context.Context) error {
```

Comment:
I'm curious how much of an improvement you saw from running the message decoding on a separate goroutine? I've tried things like this before and haven't been able to get much of a performance increase, possibly due to extra GC + channel overhead, and it increases memory usage and OOM potential if documents are very large, particularly if the `decodeCh` is buffered.

Reply:
I'll do another round of testing with and without the `decodeCh` to see if there is still an improvement in performance, previously I had noticed there was too much time wasted on I/O blocking CPU processes, so I moved decode to its own worker. The decoding process is one of the most CPU-intensive parts of the program (parsing sql is CPU intensive). I'll see if that still applies after I changed how `receiveMessages` works

Reply:
I did some tests with two versions of the connector, one with `decodeCh` and one without.

The one with the channel shows ~10% better performance on 1 million documents (500MB) in a single SCN span (one run of `receiveMessages`). However when working on larger data in a single SCN, the `decodeCh` version does indeed OOM while the normal version doesn't.

I think not OOMing is more important than the 10% performance improvement for now. Will have to optimise CPU utilisation of this connector at some stage and will address performance then. For now I'm going to revert back the `decodeCh` changes 👍🏽